### PR TITLE
New: [AEA-4277] - Deploy docker container from ECR to ECS Fargate

### PR DESCRIPTION
**Pull request by kris-szlapa**
## Summary

🎫 [AEA-4277](https://nhsd-jira.digital.nhs.uk/browse/AEA-4277) Deploy docker container from ECR to ECS Fargate
🎫 [AEA-4278](https://nhsd-jira.digital.nhs.uk/browse/AEA-4278) Connect public ALB to service in ECS Fargate

- :sparkles: New Feature
- :robot: Operational or Infrastructure Change
- :warning: Potential issues that might be caused by this change

### Details

- Deploy docker container from ECR to ECS Fargate
- Connect public ALB to service in ECS Fargate

### CODE REVIEW

File: SAMtemplates/load_balancer/main.yaml [hunk 1]
```diff
@@ -25,11 +25,29 @@ Parameters:
   CommitId:
     Type: String
 
+  KmsKey:
```
Comment (anthony-nhs) [inline SAMtemplates/load_balancer/main.yaml:28]: dont think you need this as an import

Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The unnecessary parameter has been removed.

File: .github/workflows/sam_release_code.yml [hunk 2]
```diff
@@ -147,6 +147,12 @@ jobs:
           echo "Specification location: $SPEC_PATH"
           echo "SPEC_PATH=${SPEC_PATH}" >> "$GITHUB_ENV"
       
+      - name: Set ECR Repository
+        run: echo "ECR_REPOSITORY=fhir-facade-repo" >> "$GITHUB_ENV"
+
+      - name: Set Image Tag
+        run: echo "IMAGE_TAG=${{ inputs.VERSION_NUMBER }}" >> "$GITHUB_ENV"
```
Comment (MatthewPopat-NHS) [inline .github/workflows/sam_release_code.yml:154]: Can we just inline these below? I think this just makes it harder to figure out what values are being used when reading the code.

File: SAMtemplates/load_balancer/main.yaml [hunk 3]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+        OperatingSystemFamily: LINUX
+
+  # ECS Service Configuration
+  ECSService:
+    Type: AWS::ECS::Service
+    DependsOn: ListenerHTTPS
+    Properties:
+      Cluster: !Ref ECSCluster
+      ServiceName: !Sub "${StackName}-service"
+      DesiredCount: 3
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:223]: The current defaults are 2 for prod and ref, 1 for everything else. We should probably match that for now.

File: SAMtemplates/load_balancer/main.yaml [hunk 4]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+            - !Ref ContainerSecurityGroup
+          Subnets: !Split [",", !ImportValue vpc-resources:PrivateSubnets]
+      TaskDefinition: !Ref ECSTaskDefinition
+      LoadBalancers:
+        - ContainerName: !Sub "${StackName}-container"
+          ContainerPort: !Ref ContainerPort
+          TargetGroupArn: !Ref TargetGroup
+
+  # ECS Service Auto Scaling Role
+  ECSServiceAutoScalingRole:
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:238]: We should add conditions to all the autoscaling resources so they only deploy in ref and prod.

File: SAMtemplates/load_balancer/main.yaml [hunk 5]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceAutoscaleRole
+      RoleName: !Sub "${StackName}-ecsAutoScalingRole"
+
+  # ECS Service Auto Scaling Target
+  ECSServiceAutoScalingTarget:
+    Type: AWS::ApplicationAutoScaling::ScalableTarget
+    DependsOn: ECSService
+    Properties:
+      MinCapacity: 3
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:257]: This is currently 2 in the azure pipelines so should probably match that.

File: SAMtemplates/load_balancer/main.yaml [hunk 6]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+      MaxCapacity: 4
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:258]: This is currently 10 in the azure pipelines so should probably match that.

File: SAMtemplates/load_balancer/main.yaml [hunk 7]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+    Properties:
+      PolicyName: !Sub "${StackName}-ecsAutoScalingPolicy"
+      PolicyType: TargetTrackingScaling
+      ScalingTargetId: !Ref ECSServiceAutoScalingTarget
+      TargetTrackingScalingPolicyConfiguration:
+        PredefinedMetricSpecification:
+          PredefinedMetricType: ECSServiceAverageCPUUtilization
+        ScaleInCooldown: 60
+        ScaleOutCooldown: 60
+        TargetValue: 50.0
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:276]: This is currently 70 in the azure pipelines so should probably match that.

File: SAMtemplates/load_balancer/main.yaml [hunk 8]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+  ECSServiceAutoScalingPolicy:
+    Type: AWS::ApplicationAutoScaling::ScalingPolicy
+    Properties:
+      PolicyName: !Sub "${StackName}-ecsAutoScalingPolicy"
+      PolicyType: TargetTrackingScaling
+      ScalingTargetId: !Ref ECSServiceAutoScalingTarget
+      TargetTrackingScalingPolicyConfiguration:
+        PredefinedMetricSpecification:
+          PredefinedMetricType: ECSServiceAverageCPUUtilization
+        ScaleInCooldown: 60
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:274]: This is currently 300 in the azure pipelines so should probably match that.

File: SAMtemplates/load_balancer/main.yaml [hunk 9]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+              awslogs-stream-prefix: ecs
+      RuntimePlatform:
+        CpuArchitecture: X86_64
+        OperatingSystemFamily: LINUX
+
+  # ECS Service Configuration
+  ECSService:
+    Type: AWS::ECS::Service
+    DependsOn: ListenerHTTPS
+    Properties:
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:220]: Probably need `HealthCheckGracePeriodSeconds: 300` to match the azure pipelines.

File: SAMtemplates/load_balancer/main.yaml [hunk 10]
```diff
@@ -137,6 +151,157 @@ Resources:
...
+      GroupDescription: Security group for ECS containers
+      SecurityGroupIngress:
+        - IpProtocol: tcp
+          FromPort: !Ref ContainerPort
+          ToPort: !Ref ContainerPort
+          SourceSecurityGroupId: !Ref LoadBalancerSecurityGroup
+        - IpProtocol: tcp
+          FromPort: !Ref ContainerPortValidator
+          ToPort: !Ref ContainerPortValidator
+          SourceSecurityGroupId: !Ref LoadBalancerSecurityGroup
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:300]: This shouldn't be needed, only the main fhir facade container should talk to validator, the load balancer shouldn't talk to it directly.

File: .github/workflows/sam_release_code.yml (hunk 2 above)
Comment (kris-szlapa) [reply .github/workflows/sam_release_code.yml]: The values for `fhir-facade-repo` and `validator-repo` are used directly in the `ECSTaskDefinition` resource.

File: SAMtemplates/load_balancer/main.yaml [hunk 11]
```diff
@@ -137,6 +155,184 @@ Resources:
...
+      MinCapacity:
+        !If
+          - IsRefOrProd
+          - 2
+          - 1
+      MaxCapacity:
+        !If
+          - IsRefOrProd
+          - 10
+          - 6
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:276]: no need for these conditions as the resource isn't even there for non ref and prod

File: SAMtemplates/load_balancer/main.yaml [hunk 12]
```diff
@@ -137,6 +155,184 @@ Resources:
...
+        ScaleOutCooldown:
+          !If
+            - IsRefOrProd
+            - 300
+            - 60
+        TargetValue:
+          !If
+            - IsRefOrProd
+            - 70.0
+            - 50.0
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:307]: no need for these conditions as the resource isn't even there for non ref and prod

File: SAMtemplates/load_balancer/main.yaml [hunk 13]
```diff
@@ -137,6 +155,184 @@ Resources:
...
+          PredefinedMetricType: ECSServiceAverageCPUUtilization
+        ScaleInCooldown:
+          !If
+            - IsRefOrProd
+            - 300
+            - 60
+        ScaleOutCooldown:
+          !If
+            - IsRefOrProd
+            - 300
```
Comment (MatthewPopat-NHS) [inline SAMtemplates/load_balancer/main.yaml:301]: scale out should still be 60, only scale in should be 300

File: SAMtemplates/load_balancer/main.yaml (hunk 5 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The value of `MinCapacity` has been changed to 2.

File: SAMtemplates/load_balancer/main.yaml (hunk 6 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The value of `MaxCapacity` has been changed to 10.

File: SAMtemplates/load_balancer/main.yaml (hunk 7 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The value of `TargetValue` has been changed to 70.

File: SAMtemplates/load_balancer/main.yaml (hunk 8 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The value of `ScaleInCooldown` has been changed to 300.

File: SAMtemplates/load_balancer/main.yaml (hunk 9 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml:226]: The `HealthCheckGracePeriodSeconds` property, set to `300`, has been added.

File: SAMtemplates/load_balancer/main.yaml (hunk 12 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The unnecessary condition has been removed.

File: SAMtemplates/load_balancer/main.yaml (hunk 13 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The unnecessary condition has been removed, and `ScaleOutCooldown` has been set to `60`.

File: SAMtemplates/load_balancer/main.yaml (hunk 10 above)
Comment (kris-szlapa) [reply SAMtemplates/load_balancer/main.yaml]: The direct access from the Load Balancer to the Validator container has been removed.

**Review by anthony-nhs — APPROVED** [review]

**Pull request merged**